lib/writeGltf.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+var fs = require('fs');
+var path = require('path');
+var Cesium = require('cesium');
+var defined = Cesium.defined;
+var dataUri = require('datauri');
+var async = require('async');
+var mkdirp = require('mkdirp');
+
+module.exports = writeGltf;
+
+function writeGltf(gltf, outputPath, isEmbedded, createDirectory, callback) {
+    //Create the output directory if specified
+    if (createDirectory) {
+        outputPath = path.join(path.dirname(outputPath), 'output', path.basename(outputPath));
+        mkdirp.sync(path.dirname(outputPath));
+    }
+    
+    var shaders = gltf.shaders;
+    //Write out the source objects, followed by the entire gltf object
+    if (defined(shaders)) {
+        var shaderIds = Object.keys(shaders);
+        async.each(shaderIds, function(shaderId, asyncCallback) {
+            var shader = shaders[shaderId];
+            if (defined(shader.extras) && defined(shader.extras.source)) {
+                var source = shader.extras.source;
+                delete shader.extras.source;
+
+                var filePath = path.dirname(outputPath);
+                writeSource(filePath, shaderId, shader, source, isEmbedded, asyncCallback);
+            }
+        }, function(err) {
+            if (err) {
+                if (callback) {
+                        process.nextTick(function() {

```

Comment:
Exactly.  Basically, all callbacks need to be async, e.g., as if they are called after file or network IO, so a function should never call a callback synchronously; instead, they should use `process.nextTick` so it happens later in Node's runtime loop (it is kinda like faking very fast file IO).  So, in the case of decoding a data URI, we could call the callback synchronously for data URIs, but it would be a bad practice because, from the user's perspective, sometimes it would be called synchronously (data URI) and sometimes not (file url) so the Node.js best practice is to always make it async.